teachertool/src/components/SplitPane.tsx

```diff
@@ -5,20 +5,115 @@ import { classList } from "react-common/components/util";
-    defaultSize: number | string;
+    defaultSize: number | string; // The size to reset to when double clicking the splitter.
+    startingSize?: number | string; // The size to use initially when creating the splitter. Defaults to `defaultSize`.
     primary: "left" | "right";
     left: React.ReactNode;
     right: React.ReactNode;
+    leftMinSize?: number | string;
+    rightMinSize?: number | string;
+    onResizeEnd?: (size: number | string) => void;
 }
 
-export const SplitPane: React.FC<IProps> = ({ className, split, left, right }) => {
+export const SplitPane: React.FC<IProps> = ({
+    className,
+    split,
+    defaultSize,
+    startingSize,
+    left,
+    right,
+    leftMinSize,
+    rightMinSize,
+    onResizeEnd,
+}) => {
+    const [size, setSize] = React.useState(startingSize ?? defaultSize);
+    const [isResizing, setIsResizing] = React.useState(false);
+    const containerRef = React.useRef<HTMLDivElement>(null);
+
+    React.useEffect(() => {
+        if (!isResizing) {
+            onResizeEnd?.(size);
+        }
+    }, [isResizing]);
+
+    function handleResizeMouse(event: MouseEvent) {
+        handleResize(event.clientX, event.clientY);
+    }
+
+    function handleResizeTouch(event: TouchEvent) {
+        if (event.touches.length >= 1) {
+            handleResize(event.touches[0].clientX, event.touches[0].clientY);
+        }
+    }
+
+    function handleResize(clientX: number, clientY: number) {
+        const containerRect = containerRef.current?.getBoundingClientRect();
+        if (containerRect) {
+            const newSize =
+                split === "vertical"
+                    ? `${((clientX - containerRect.left) / containerRect.width) * 100}%`
+                    : `${((clientY - containerRect.top) / containerRect.height) * 100}%`;
+
+            setIsResizing(true); // Do this here rather than inside startResizing to prevent interference with double click detection.
+            setSize(newSize);
+        }
+    }
+
+    function startResizing(event: React.MouseEvent | React.TouchEvent) {
+        event.preventDefault();
+        document.addEventListener("mousemove", handleResizeMouse);
+        document.addEventListener("mouseup", endResizing);
+        document.addEventListener("touchmove", handleResizeTouch);
+        document.addEventListener("touchend", endResizing);
+    }
+
+    function endResizing() {
+        document.removeEventListener("mousemove", handleResizeMouse);
+        document.removeEventListener("mouseup", endResizing);
+        document.removeEventListener("touchmove", handleResizeTouch);
+        document.removeEventListener("touchend", endResizing);
+
+        setIsResizing(false);
+    }
+
+    function setToDefaultSize() {
+        setSize(defaultSize);
+        onResizeEnd?.(defaultSize);
+    }
+
+    const leftStyle: React.CSSProperties = { flexBasis: size };
+    if (split === "vertical") {
+        leftStyle.minWidth = leftMinSize;
```

Comment:
Do we need to make the `leftMinSize` and `rightMinSize` props required? If they aren't set, I think these calculations will fail sine we don't have default values for `leftMinSize` and `rightMinSize`.

Reply:
Hmm, I don't think we should allow them to be fully unset, since it's possible to get into a state where you the splitter goes offscreen or something and then you can never get out of that state, even when you refresh the page.

We could have a default value, but when in doubt, I think it's simpler to just have the caller tell us what these values should be. The caller will always have more context than this component does, and it's not too burdensome to include.

Reply:
Yes, I agree, but right now these props are optional: https://github.com/microsoft/pxt/pull/9887/files#diff-5c44e5a8b91e45e13e21ff9f1a12da3634d9282699fb7a757190de64abcff78dR13-R14. I think they should just be required props to avoid possible problems.